From 0aae4a6a4810e4d6f2c20b9b9868d2691681da05 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20G=C3=B3rny?= <mgorny@gentoo.org>
Date: Tue, 16 Jun 2020 20:43:55 +0200
Subject: [PATCH] Prevent linking to duplicate .a libs and dylib

Fix various tool libraries not to link to clang's .a libraries and dylib
simultaneously.  This may cause breakage, in particular through
duplicate command-line option declarations.

Differential Revision: https://reviews.llvm.org/D81967
---
 clang-apply-replacements/CMakeLists.txt            |  4 +++-
 clang-change-namespace/CMakeLists.txt              |  4 +++-
 clang-doc/CMakeLists.txt                           |  4 +++-
 clang-include-fixer/CMakeLists.txt                 |  6 +++++-
 .../find-all-symbols/CMakeLists.txt                |  4 +++-
 clang-move/CMakeLists.txt                          |  4 +++-
 clang-query/CMakeLists.txt                         |  4 +++-
 clang-reorder-fields/CMakeLists.txt                |  4 +++-
 clang-tidy/CMakeLists.txt                          |  7 +++++--
 clang-tidy/abseil/CMakeLists.txt                   |  8 ++++++--
 clang-tidy/android/CMakeLists.txt                  |  8 ++++++--
 clang-tidy/boost/CMakeLists.txt                    |  8 ++++++--
 clang-tidy/bugprone/CMakeLists.txt                 | 10 +++++++---
 clang-tidy/cert/CMakeLists.txt                     | 12 ++++++++----
 clang-tidy/cppcoreguidelines/CMakeLists.txt        | 14 +++++++++-----
 clang-tidy/darwin/CMakeLists.txt                   |  8 ++++++--
 clang-tidy/fuchsia/CMakeLists.txt                  | 10 +++++++---
 clang-tidy/google/CMakeLists.txt                   | 10 +++++++---
 clang-tidy/hicpp/CMakeLists.txt                    | 14 +++++++++-----
 clang-tidy/linuxkernel/CMakeLists.txt              |  8 ++++++--
 clang-tidy/llvm/CMakeLists.txt                     | 10 +++++++---
 clang-tidy/misc/CMakeLists.txt                     |  8 ++++++--
 clang-tidy/modernize/CMakeLists.txt                | 10 +++++++---
 clang-tidy/mpi/CMakeLists.txt                      |  8 ++++++--
 clang-tidy/objc/CMakeLists.txt                     |  8 ++++++--
 clang-tidy/openmp/CMakeLists.txt                   |  8 ++++++--
 clang-tidy/performance/CMakeLists.txt              |  8 ++++++--
 clang-tidy/plugin/CMakeLists.txt                   |  8 ++++++--
 clang-tidy/portability/CMakeLists.txt              |  8 ++++++--
 clang-tidy/readability/CMakeLists.txt              |  8 ++++++--
 clang-tidy/utils/CMakeLists.txt                    |  6 +++++-
 clang-tidy/zircon/CMakeLists.txt                   |  8 ++++++--
 clangd/CMakeLists.txt                              | 12 ++++++++----
 clangd/unittests/CMakeLists.txt                    |  1 -
 34 files changed, 189 insertions(+), 73 deletions(-)

diff --git a/clang-apply-replacements/CMakeLists.txt b/clang-apply-replacements/CMakeLists.txt
index 5bfdcb4..27383b4 100644
--- a/clang-apply-replacements/CMakeLists.txt
+++ b/clang-apply-replacements/CMakeLists.txt
@@ -4,8 +4,10 @@ set(LLVM_LINK_COMPONENTS
 
 add_clang_library(clangApplyReplacements
   lib/Tooling/ApplyReplacements.cpp
+)
 
-  LINK_LIBS
+clang_target_link_libraries(clangApplyReplacements
+  PRIVATE
   clangAST
   clangBasic
   clangRewrite
diff --git a/clang-change-namespace/CMakeLists.txt b/clang-change-namespace/CMakeLists.txt
index 1783064..dccfd7c 100644
--- a/clang-change-namespace/CMakeLists.txt
+++ b/clang-change-namespace/CMakeLists.txt
@@ -4,8 +4,10 @@ set(LLVM_LINK_COMPONENTS
 
 add_clang_library(clangChangeNamespace
   ChangeNamespace.cpp
+)
 
-  LINK_LIBS
+clang_target_link_libraries(clangChangeNamespace
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
diff --git a/clang-doc/CMakeLists.txt b/clang-doc/CMakeLists.txt
index c301ad5..6f303a6 100644
--- a/clang-doc/CMakeLists.txt
+++ b/clang-doc/CMakeLists.txt
@@ -14,8 +14,10 @@ add_clang_library(clangDoc
   Representation.cpp
   Serialize.cpp
   YAMLGenerator.cpp
+)
 
-  LINK_LIBS
+clang_target_link_libraries(clangDoc
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
diff --git a/clang-include-fixer/CMakeLists.txt b/clang-include-fixer/CMakeLists.txt
index f27f740..d8685cb 100644
--- a/clang-include-fixer/CMakeLists.txt
+++ b/clang-include-fixer/CMakeLists.txt
@@ -11,6 +11,11 @@ add_clang_library(clangIncludeFixer
   YamlSymbolIndex.cpp
 
   LINK_LIBS
+  findAllSymbols
+  )
+
+clang_target_link_libraries(clangIncludeFixer
+  PRIVATE
   clangAST
   clangBasic
   clangFormat
@@ -21,7 +26,6 @@ add_clang_library(clangIncludeFixer
   clangSerialization
   clangTooling
   clangToolingCore
-  findAllSymbols
   )
 
 add_subdirectory(plugin)
diff --git a/clang-include-fixer/find-all-symbols/CMakeLists.txt b/clang-include-fixer/find-all-symbols/CMakeLists.txt
index c5fe19b..06a2324 100644
--- a/clang-include-fixer/find-all-symbols/CMakeLists.txt
+++ b/clang-include-fixer/find-all-symbols/CMakeLists.txt
@@ -11,8 +11,10 @@ add_clang_library(findAllSymbols
   PragmaCommentHandler.cpp
   STLPostfixHeaderMap.cpp
   SymbolInfo.cpp
+  )
 
-  LINK_LIBS
+clang_target_link_libraries(findAllSymbols
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
diff --git a/clang-move/CMakeLists.txt b/clang-move/CMakeLists.txt
index c63127e..9ea4099 100644
--- a/clang-move/CMakeLists.txt
+++ b/clang-move/CMakeLists.txt
@@ -5,8 +5,10 @@ set(LLVM_LINK_COMPONENTS
 add_clang_library(clangMove
   Move.cpp
   HelperDeclRefGraph.cpp
+  )
 
-  LINK_LIBS
+clang_target_link_libraries(clangMove
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
diff --git a/clang-query/CMakeLists.txt b/clang-query/CMakeLists.txt
index d1d68d5..bc76bae 100644
--- a/clang-query/CMakeLists.txt
+++ b/clang-query/CMakeLists.txt
@@ -6,8 +6,10 @@ set(LLVM_LINK_COMPONENTS
 add_clang_library(clangQuery
   Query.cpp
   QueryParser.cpp
+  )
 
-  LINK_LIBS
+clang_target_link_libraries(clangQuery
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
diff --git a/clang-reorder-fields/CMakeLists.txt b/clang-reorder-fields/CMakeLists.txt
index 9c75d78..e555fd5 100644
--- a/clang-reorder-fields/CMakeLists.txt
+++ b/clang-reorder-fields/CMakeLists.txt
@@ -2,8 +2,10 @@ set(LLVM_LINK_COMPONENTS support)
 
 add_clang_library(clangReorderFields
   ReorderFieldsAction.cpp
+)
 
-  LINK_LIBS
+clang_target_link_libraries(clangReorderFields
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
diff --git a/clang-tidy/CMakeLists.txt b/clang-tidy/CMakeLists.txt
index 6dadb27..329df19 100644
--- a/clang-tidy/CMakeLists.txt
+++ b/clang-tidy/CMakeLists.txt
@@ -14,8 +14,10 @@ add_clang_library(clangTidy
 
   DEPENDS
   ClangSACheckers
+  )
 
-  LINK_LIBS
+clang_target_link_libraries(clangTidy
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
@@ -31,7 +33,8 @@ add_clang_library(clangTidy
   )
 
 if(CLANG_ENABLE_STATIC_ANALYZER)
-  target_link_libraries(clangTidy PRIVATE
+  clang_target_link_libraries(clangTidy
+    PRIVATE
     clangStaticAnalyzerCore
     clangStaticAnalyzerFrontend
   )
diff --git a/clang-tidy/abseil/CMakeLists.txt b/clang-tidy/abseil/CMakeLists.txt
index 3f88da6..0f040e5 100644
--- a/clang-tidy/abseil/CMakeLists.txt
+++ b/clang-tidy/abseil/CMakeLists.txt
@@ -22,11 +22,15 @@ add_clang_library(clangTidyAbseilModule
   UpgradeDurationConversionsCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyAbseilModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/android/CMakeLists.txt b/clang-tidy/android/CMakeLists.txt
index 9d04003..d29adc6 100644
--- a/clang-tidy/android/CMakeLists.txt
+++ b/clang-tidy/android/CMakeLists.txt
@@ -20,10 +20,14 @@ add_clang_library(clangTidyAndroidModule
   ComparisonInTempFailureRetryCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyAndroidModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/boost/CMakeLists.txt b/clang-tidy/boost/CMakeLists.txt
index 059f6e9..481a709 100644
--- a/clang-tidy/boost/CMakeLists.txt
+++ b/clang-tidy/boost/CMakeLists.txt
@@ -5,10 +5,14 @@ add_clang_library(clangTidyBoostModule
   UseToStringCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyBoostModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/bugprone/CMakeLists.txt b/clang-tidy/bugprone/CMakeLists.txt
index 2cb28f4..d664f5e 100644
--- a/clang-tidy/bugprone/CMakeLists.txt
+++ b/clang-tidy/bugprone/CMakeLists.txt
@@ -52,13 +52,17 @@ add_clang_library(clangTidyBugproneModule
   VirtualNearMissCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyCppCoreGuidelinesModule
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyBugproneModule
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyCppCoreGuidelinesModule
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/cert/CMakeLists.txt b/clang-tidy/cert/CMakeLists.txt
index 66ea2a1..75a1334 100644
--- a/clang-tidy/cert/CMakeLists.txt
+++ b/clang-tidy/cert/CMakeLists.txt
@@ -17,10 +17,6 @@ add_clang_library(clangTidyCERTModule
   VariadicFunctionDefCheck.cpp
 
   LINK_LIBS
-  clangAST
-  clangASTMatchers
-  clangBasic
-  clangLex
   clangTidy
   clangTidyBugproneModule
   clangTidyGoogleModule
@@ -29,3 +25,11 @@ add_clang_library(clangTidyCERTModule
   clangTidyReadabilityModule
   clangTidyUtils
   )
+
+clang_target_link_libraries(clangTidyCERTModule
+  PRIVATE
+  clangAST
+  clangASTMatchers
+  clangBasic
+  clangLex
+  )
diff --git a/clang-tidy/cppcoreguidelines/CMakeLists.txt b/clang-tidy/cppcoreguidelines/CMakeLists.txt
index 13c15bc..e48bc1c 100644
--- a/clang-tidy/cppcoreguidelines/CMakeLists.txt
+++ b/clang-tidy/cppcoreguidelines/CMakeLists.txt
@@ -23,15 +23,19 @@ add_clang_library(clangTidyCppCoreGuidelinesModule
   SpecialMemberFunctionsCheck.cpp
 
   LINK_LIBS
-  clangAST
-  clangASTMatchers
-  clangBasic
-  clangLex
-  clangSerialization
   clangTidy
   clangTidyMiscModule
   clangTidyModernizeModule
   clangTidyReadabilityModule
   clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyCppCoreGuidelinesModule
+  PRIVATE
+  clangAST
+  clangASTMatchers
+  clangBasic
+  clangLex
+  clangSerialization
   clangTooling
   )
diff --git a/clang-tidy/darwin/CMakeLists.txt b/clang-tidy/darwin/CMakeLists.txt
index c650efb..feec79c 100644
--- a/clang-tidy/darwin/CMakeLists.txt
+++ b/clang-tidy/darwin/CMakeLists.txt
@@ -6,11 +6,15 @@ add_clang_library(clangTidyDarwinModule
   DispatchOnceNonstaticCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyDarwinModule
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/fuchsia/CMakeLists.txt b/clang-tidy/fuchsia/CMakeLists.txt
index 30b319e..0148b20 100644
--- a/clang-tidy/fuchsia/CMakeLists.txt
+++ b/clang-tidy/fuchsia/CMakeLists.txt
@@ -12,11 +12,15 @@ add_clang_library(clangTidyFuchsiaModule
   VirtualInheritanceCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyGoogleModule
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyFuchsiaModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyGoogleModule
-  clangTidyUtils
   )
diff --git a/clang-tidy/google/CMakeLists.txt b/clang-tidy/google/CMakeLists.txt
index 0836893..302561b 100644
--- a/clang-tidy/google/CMakeLists.txt
+++ b/clang-tidy/google/CMakeLists.txt
@@ -21,11 +21,15 @@ add_clang_library(clangTidyGoogleModule
   UsingNamespaceDirectiveCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyReadabilityModule
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyGoogleModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyReadabilityModule
-  clangTidyUtils
   )
diff --git a/clang-tidy/hicpp/CMakeLists.txt b/clang-tidy/hicpp/CMakeLists.txt
index 4cf2676..b4c9eb7 100644
--- a/clang-tidy/hicpp/CMakeLists.txt
+++ b/clang-tidy/hicpp/CMakeLists.txt
@@ -8,11 +8,6 @@ add_clang_library(clangTidyHICPPModule
   SignedBitwiseCheck.cpp
 
   LINK_LIBS
-  clangAST
-  clangASTMatchers
-  clangBasic
-  clangLex
-  clangSerialization
   clangTidy
   clangTidyBugproneModule
   clangTidyCppCoreGuidelinesModule
@@ -23,3 +18,12 @@ add_clang_library(clangTidyHICPPModule
   clangTidyReadabilityModule
   clangTidyUtils
   )
+
+clang_target_link_libraries(clangTidyHICPPModule
+  PRIVATE
+  clangAST
+  clangASTMatchers
+  clangBasic
+  clangLex
+  clangSerialization
+  )
diff --git a/clang-tidy/linuxkernel/CMakeLists.txt b/clang-tidy/linuxkernel/CMakeLists.txt
index f0e766d..bfcc2ba 100644
--- a/clang-tidy/linuxkernel/CMakeLists.txt
+++ b/clang-tidy/linuxkernel/CMakeLists.txt
@@ -5,10 +5,14 @@ add_clang_library(clangTidyLinuxKernelModule
   MustCheckErrsCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyLinuxKernelModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/llvm/CMakeLists.txt b/clang-tidy/llvm/CMakeLists.txt
index c035596..bff128c 100644
--- a/clang-tidy/llvm/CMakeLists.txt
+++ b/clang-tidy/llvm/CMakeLists.txt
@@ -9,12 +9,16 @@ add_clang_library(clangTidyLLVMModule
   TwineLocalCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyReadabilityModule
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyLLVMModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyReadabilityModule
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/misc/CMakeLists.txt b/clang-tidy/misc/CMakeLists.txt
index 3fc1521..37bc2aa 100644
--- a/clang-tidy/misc/CMakeLists.txt
+++ b/clang-tidy/misc/CMakeLists.txt
@@ -17,13 +17,17 @@ add_clang_library(clangTidyMiscModule
   UnusedUsingDeclsCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyMiscModule
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
   clangSerialization
-  clangTidy
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/modernize/CMakeLists.txt b/clang-tidy/modernize/CMakeLists.txt
index 36193f0..162f445 100644
--- a/clang-tidy/modernize/CMakeLists.txt
+++ b/clang-tidy/modernize/CMakeLists.txt
@@ -36,12 +36,16 @@ add_clang_library(clangTidyModernizeModule
   UseUsingCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyReadabilityModule
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyModernizeModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyReadabilityModule
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/mpi/CMakeLists.txt b/clang-tidy/mpi/CMakeLists.txt
index 5be7b36..b36767a 100644
--- a/clang-tidy/mpi/CMakeLists.txt
+++ b/clang-tidy/mpi/CMakeLists.txt
@@ -6,13 +6,17 @@ add_clang_library(clangTidyMPIModule
   TypeMismatchCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyMPIModule
+  PRIVATE
   clangAnalysis
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   clangTooling
   clangStaticAnalyzerCheckers
   )
diff --git a/clang-tidy/objc/CMakeLists.txt b/clang-tidy/objc/CMakeLists.txt
index 68dda65..3624d2f 100644
--- a/clang-tidy/objc/CMakeLists.txt
+++ b/clang-tidy/objc/CMakeLists.txt
@@ -9,10 +9,14 @@ add_clang_library(clangTidyObjCModule
   SuperSelfCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyObjCModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/openmp/CMakeLists.txt b/clang-tidy/openmp/CMakeLists.txt
index af95704..d182b29 100644
--- a/clang-tidy/openmp/CMakeLists.txt
+++ b/clang-tidy/openmp/CMakeLists.txt
@@ -8,9 +8,13 @@ add_clang_library(clangTidyOpenMPModule
   UseDefaultNoneCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyOpenMPModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/performance/CMakeLists.txt b/clang-tidy/performance/CMakeLists.txt
index d1f9897..c9a7e26 100644
--- a/clang-tidy/performance/CMakeLists.txt
+++ b/clang-tidy/performance/CMakeLists.txt
@@ -18,11 +18,15 @@ add_clang_library(clangTidyPerformanceModule
   UnnecessaryValueParamCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyPerformanceModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangAnalysis
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clang-tidy/plugin/CMakeLists.txt b/clang-tidy/plugin/CMakeLists.txt
index 4adc3f2..0bfe122 100644
--- a/clang-tidy/plugin/CMakeLists.txt
+++ b/clang-tidy/plugin/CMakeLists.txt
@@ -2,12 +2,16 @@ add_clang_library(clangTidyPlugin
   ClangTidyPlugin.cpp
 
   LINK_LIBS
+  clangTidy
+  ${ALL_CLANG_TIDY_CHECKS}
+  )
+
+clang_target_link_libraries(clangTidyPlugin
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangFrontend
   clangSema
-  clangTidy
   clangTooling
-  ${ALL_CLANG_TIDY_CHECKS}
   )
diff --git a/clang-tidy/portability/CMakeLists.txt b/clang-tidy/portability/CMakeLists.txt
index 0420a18..bbe9ccb 100644
--- a/clang-tidy/portability/CMakeLists.txt
+++ b/clang-tidy/portability/CMakeLists.txt
@@ -5,11 +5,15 @@ add_clang_library(clangTidyPortabilityModule
   SIMDIntrinsicsCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyPortabilityModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/readability/CMakeLists.txt b/clang-tidy/readability/CMakeLists.txt
index 97144af..5af900d 100644
--- a/clang-tidy/readability/CMakeLists.txt
+++ b/clang-tidy/readability/CMakeLists.txt
@@ -41,11 +41,15 @@ add_clang_library(clangTidyReadabilityModule
   UppercaseLiteralSuffixCheck.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyReadabilityModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   clangTooling
   )
diff --git a/clang-tidy/utils/CMakeLists.txt b/clang-tidy/utils/CMakeLists.txt
index fc383a3..5c837ca 100644
--- a/clang-tidy/utils/CMakeLists.txt
+++ b/clang-tidy/utils/CMakeLists.txt
@@ -18,11 +18,15 @@ add_clang_library(clangTidyUtils
   UsingInserter.cpp
 
   LINK_LIBS
+  clangTidy
+  )
+
+clang_target_link_libraries(clangTidyUtils
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
   clangSema
-  clangTidy
   clangTransformer
   )
diff --git a/clang-tidy/zircon/CMakeLists.txt b/clang-tidy/zircon/CMakeLists.txt
index 7aa7cd3..71db5e9 100644
--- a/clang-tidy/zircon/CMakeLists.txt
+++ b/clang-tidy/zircon/CMakeLists.txt
@@ -5,10 +5,14 @@ add_clang_library(clangTidyZirconModule
   ZirconTidyModule.cpp
 
   LINK_LIBS
+  clangTidy
+  clangTidyUtils
+  )
+
+clang_target_link_libraries(clangTidyZirconModule
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangLex
-  clangTidy
-  clangTidyUtils
   )
diff --git a/clangd/CMakeLists.txt b/clangd/CMakeLists.txt
index e3eccb5..c797c4e 100644
--- a/clangd/CMakeLists.txt
+++ b/clangd/CMakeLists.txt
@@ -110,6 +110,14 @@ add_clang_library(clangDaemon
   refactor/Tweak.cpp
 
   LINK_LIBS
+  clangTidy
+  ${LLVM_PTHREAD_LIB}
+  ${CLANGD_ATOMIC_LIB}
+  ${ALL_CLANG_TIDY_CHECKS}
+  )
+
+clang_target_link_libraries(clangDaemon
+  PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
@@ -120,15 +128,11 @@ add_clang_library(clangDaemon
   clangLex
   clangSema
   clangSerialization
-  clangTidy
   clangTooling
   clangToolingCore
   clangToolingInclusions
   clangToolingRefactoring
   clangToolingSyntax
-  ${LLVM_PTHREAD_LIB}
-  ${CLANGD_ATOMIC_LIB}
-  ${ALL_CLANG_TIDY_CHECKS}
   )
 
 add_subdirectory(refactor/tweaks)
diff --git a/clangd/unittests/CMakeLists.txt b/clangd/unittests/CMakeLists.txt
index 62113c6..6c2b56c 100644
--- a/clangd/unittests/CMakeLists.txt
+++ b/clangd/unittests/CMakeLists.txt
@@ -103,7 +103,6 @@ target_link_libraries(ClangdTests
   PRIVATE
   clangDaemon
   clangTidy
-  LLVMSupport
   LLVMTestingSupport
   )
 
-- 
2.27.0.windows.1

